File: go/test/moon_direct_test.go
package sdktest

import (
	"encoding/json"
	"os"
	"strings"
	"testing"

	sdk "voxgigsolardemosdk"
	"voxgigsolardemosdk/core"
)

func TestMoonDirect(t *testing.T) {
	t.Run("direct-list-moon", func(t *testing.T) {
		setup := moonDirectSetup([]any{
			map[string]any{"id": "direct01"},
			map[string]any{"id": "direct02"},
		})
		client := setup.client

		params := map[string]any{}
		if setup.live {
			params["planet_id"] = setup.idmap["planet01"]
		} else {
			params["planet_id"] = "direct01"
		}

		result, err := client.Direct(map[string]any{
			"path":   "api/planet/{planet_id}/moon",
			"method": "GET",
			"params": params,
		})
		if err != nil {
			t.Fatalf("direct failed: %v", err)
		}

		if result["ok"] != true {
			t.Fatalf("expected ok to be true, got %v", result["ok"])
		}
		if core.ToInt(result["status"]) != 200 {
			t.Fatalf("expected status 200, got %v", result["status"])
		}

		if !setup.live {
			if dataList, ok := result["data"].([]any); ok {
				if len(dataList) != 2 {
					t.Fatalf("expected 2 items, got %d", len(dataList))
				}
			} else {
				t.Fatalf("expected data to be an array, got %T", result["data"])
			}

			if len(*setup.calls) != 1 {
				t.Fatalf("expected 1 call, got %d", len(*setup.calls))
			}
			call := (*setup.calls)[0]
			if initMap, ok := call["init"].(map[string]any); ok {
				if initMap["method"] != "GET" {
					t.Fatalf("expected method GET, got %v", initMap["method"])
				}
			}
			if url, ok := call["url"].(string); ok {
				if !strings.Contains(url, "direct01") {
					t.Fatalf("expected url to contain direct01, got %v", url)
				}
			}
		}
	})

	t.Run("direct-load-moon", func(t *testing.T) {
		setup := moonDirectSetup(map[string]any{"id": "direct01"})
		client := setup.client

		params := map[string]any{}
		if setup.live {
			listParams := map[string]any{}
			listParams["planet_id"] = setup.idmap["planet01"]
			listResult, listErr := client.Direct(map[string]any{
				"path":   "api/planet/{planet_id}/moon",
				"method": "GET",
				"params": listParams,
			})
			if listErr != nil {
				t.Fatalf("list for load setup failed: %v", listErr)
			}
			if listResult["ok"] != true {
				t.Fatalf("list for load setup not ok: %v", listResult)
			}

			// Get first entity ID from list
			listData, _ := listResult["data"].([]any)
			if len(listData) == 0 {
				t.Skip("no entities to load in live mode")
			}
			firstEnt := core.ToMapAny(listData[0])
			params["id"] = firstEnt["id"]
			params["planet_id"] = setup.idmap["planet01"]
		} else {
			params["id"] = "direct01"
			params["planet_id"] = "direct02"
		}

		result, err := client.Direct(map[string]any{
			"path":   "api/planet/{planet_id}/moon/{id}",
			"method": "GET",
			"params": params,
		})
		if err != nil {
			t.Fatalf("direct failed: %v", err)
		}

		if result["ok"] != true {
			t.Fatalf("expected ok to be true, got %v", result["ok"])
		}
		if core.ToInt(result["status"]) != 200 {
			t.Fatalf("expected status 200, got %v", result["status"])
		}
		if result["data"] == nil {
			t.Fatal("expected data to be non-nil")
		}

		if !setup.live {
			if dataMap, ok := result["data"].(map[string]any); ok {
				if dataMap["id"] != "direct01" {
					t.Fatalf("expected data.id to be direct01, got %v", dataMap["id"])
				}
			}

			if len(*setup.calls) != 1 {
				t.Fatalf("expected 1 call, got %d", len(*setup.calls))
			}
			call := (*setup.calls)[0]
			if initMap, ok := call["init"].(map[string]any); ok {
				if initMap["method"] != "GET" {
					t.Fatalf("expected method GET, got %v", initMap["method"])
				}
			}
			if url, ok := call["url"].(string); ok {
				if !strings.Contains(url, "direct01") {
					t.Fatalf("expected url to contain direct01, got %v", url)
				}
				if !strings.Contains(url, "direct02") {
					t.Fatalf("expected url to contain direct02, got %v", url)
				}
			}
		}
	})

}

type moonDirectSetupResult struct {
	client *sdk.SolardemoSDK
	calls  *[]map[string]any
	live   bool
	idmap  map[string]any
}

func moonDirectSetup(mockres any) *moonDirectSetupResult {
	loadEnvLocal()

	calls := &[]map[string]any{}

	env := envOverride(map[string]any{
		"SOLARDEMO_TEST_MOON_ENTID": map[string]any{},
		"SOLARDEMO_TEST_LIVE":    "FALSE",
		"SOLARDEMO_APIKEY":       "NONE",
	})

	live := env["SOLARDEMO_TEST_LIVE"] == "TRUE"

	if live {
		mergedOpts := map[string]any{
			"apikey": env["SOLARDEMO_APIKEY"],
		}
		client := sdk.NewSolardemoSDK(mergedOpts)

		idmap := map[string]any{}
		if entidRaw, ok := env["SOLARDEMO_TEST_MOON_ENTID"]; ok {
			if entidStr, ok := entidRaw.(string); ok && strings.HasPrefix(entidStr, "{") {
				json.Unmarshal([]byte(entidStr), &idmap)
			} else if entidMap, ok := entidRaw.(map[string]any); ok {
				idmap = entidMap
			}
		}

		return &moonDirectSetupResult{client: client, calls: calls, live: true, idmap: idmap}
	}

	mockFetch := func(url string, init map[string]any) (map[string]any, error) {
		*calls = append(*calls, map[string]any{"url": url, "init": init})
		return map[string]any{
			"status":     200,
			"statusText": "OK",
			"headers":    map[string]any{},
			"json": (func() any)(func() any {
				if mockres != nil {
					return mockres
				}
				return map[string]any{"id": "direct01"}
			}),
		}, nil
	}

	client := sdk.NewSolardemoSDK(map[string]any{
		"base": "http://localhost:8080",
		"system": map[string]any{
			"fetch": (func(string, map[string]any) (map[string]any, error))(mockFetch),
		},
	})

	return &moonDirectSetupResult{client: client, calls: calls, live: false, idmap: map[string]any{}}
}

var _ = os.Getenv
var _ = json.Unmarshal

File: java/test/MoonDirectTest.java
package voxgig.solardemosdk.sdktest;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertNotNull;
import static org.junit.jupiter.api.Assertions.assertTrue;

import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.function.BiFunction;
import java.util.function.Supplier;

import org.junit.jupiter.api.Assumptions;
import org.junit.jupiter.api.Test;

import voxgig.solardemosdk.core.Helpers;
import voxgig.solardemosdk.core.SolardemoSDK;
import voxgig.solardemosdk.utility.Json;

@SuppressWarnings({"unchecked", "unused"})
public class MoonDirectTest {

  static Map<String, Object> jm(Object... kv) {
    Map<String, Object> out = new LinkedHashMap<>();
    for (int i = 0; i < kv.length - 1; i += 2) {
      out.put(String.valueOf(kv[i]), kv[i + 1]);
    }
    return out;
  }

  @Test
  public void directListMoon() {
    List<Object> mockres = new ArrayList<>();
    mockres.add(jm("id", "direct01"));
    mockres.add(jm("id", "direct02"));
    DirectSetup setup = directSetup(mockres);
    String mode = setup.live ? "live" : "unit";
    String reason = RunnerSupport.skipReason("direct", "direct-list-moon", mode);
    Assumptions.assumeTrue(reason == null,
        reason == null || "".equals(reason)
            ? "skipped via sdk-test-control.json" : reason);
    if (setup.live) {
      for (String liveKey : new String[] { "planet01" }) {
        Assumptions.assumeTrue(setup.idmap.get(liveKey) != null,
            "live test needs " + liveKey + " via *_ENTID env var (synthetic IDs only)");
      }
    }
    SolardemoSDK client = setup.client;

    Map<String, Object> params = new LinkedHashMap<>();
    if (setup.live) {
      params.put("planet_id", setup.idmap.get("planet01"));
    }
    else {
      params.put("planet_id", "direct01");
    }

    Map<String, Object> result = client.direct(jm(
        "path", "api/planet/{planet_id}/moon",
        "method", "GET",
        "params", params));
    if (setup.live) {
      // Live mode is lenient: synthetic IDs frequently 4xx and the
      // list-response shape varies wildly across public APIs. Skip
      // rather than fail when the call doesn't return a usable list.
      Assumptions.assumeTrue(Boolean.TRUE.equals(result.get("ok")),
          "list call not ok (likely synthetic IDs against live API): " + result);
      int status = Helpers.toInt(result.get("status"));
      Assumptions.assumeTrue(status >= 200 && status < 300,
          "expected 2xx status, got " + result.get("status"));
    }
    else {
      assertEquals(true, result.get("ok"), "expected ok to be true");
      assertEquals(200, Helpers.toInt(result.get("status")), "expected status 200");
    }

    if (!setup.live) {
      assertTrue(result.get("data") instanceof List,
          "expected data to be an array, got " + result.get("data"));
      assertEquals(2, ((List<Object>) result.get("data")).size(), "expected 2 items");

      assertEquals(1, setup.calls.size(), "expected 1 call");
      Map<String, Object> call = setup.calls.get(0);
      Map<String, Object> initMap = Helpers.toMapAny(call.get("init"));
      if (initMap != null) {
        assertEquals("GET", initMap.get("method"), "expected method GET");
      }
      String url = call.get("url") instanceof String ? (String) call.get("url") : "";
      assertTrue(url.contains("direct01"),
          "expected url to contain direct01, got " + url);
    }
  }

  @Test
  public void directLoadMoon() {
    DirectSetup setup = directSetup(jm("id", "direct01"));
    String mode = setup.live ? "live" : "unit";
    String reason = RunnerSupport.skipReason("direct", "direct-load-moon", mode);
    Assumptions.assumeTrue(reason == null,
        reason == null || "".equals(reason)
            ? "skipped via sdk-test-control.json" : reason);
    if (setup.live) {
      for (String liveKey : new String[] { "planet01" }) {
        Assumptions.assumeTrue(setup.idmap.get(liveKey) != null,
            "live test needs " + liveKey + " via *_ENTID env var (synthetic IDs only)");
      }
    }
    SolardemoSDK client = setup.client;

    Map<String, Object> params = new LinkedHashMap<>();
    Map<String, Object> query = new LinkedHashMap<>();
    if (setup.live) {
      Map<String, Object> listParams = new LinkedHashMap<>();
      listParams.put("planet_id", setup.idmap.get("planet01"));
      Map<String, Object> listResult = client.direct(jm(
          "path", "api/planet/{planet_id}/moon",
          "method", "GET",
          "params", listParams));
      Assumptions.assumeTrue(Boolean.TRUE.equals(listResult.get("ok")),
          "list call not ok (likely synthetic IDs against live API): " + listResult);

      // Get first entity ID from list
      List<Object> listData = listResult.get("data") instanceof List
          ? (List<Object>) listResult.get("data") : new ArrayList<>();
      Assumptions.assumeTrue(!listData.isEmpty(), "no entities to load in live mode");
      Map<String, Object> firstEnt = Helpers.toMapAny(listData.get(0));
      params.put("id", firstEnt.get("id"));
      params.put("planet_id", setup.idmap.get("planet01"));
    }
    else {
      params.put("id", "direct01");
      params.put("planet_id", "direct02");
    }

    Map<String, Object> result = client.direct(jm(
        "path", "api/planet/{planet_id}/moon/{id}",
        "method", "GET",
        "params", params,
        "query", query));
    if (setup.live) {
      // Live mode is lenient: synthetic IDs frequently 4xx. Skip rather
      // than fail when the load endpoint isn't reachable with the IDs we
      // can construct from setup.idmap.
      Assumptions.assumeTrue(Boolean.TRUE.equals(result.get("ok")),
          "load call not ok (likely synthetic IDs against live API): " + result);
      int status = Helpers.toInt(result.get("status"));
      Assumptions.assumeTrue(status >= 200 && status < 300,
          "expected 2xx status, got " + result.get("status"));
    }
    else {
      assertEquals(true, result.get("ok"), "expected ok to be true");
      assertEquals(200, Helpers.toInt(result.get("status")), "expected status 200");
      assertNotNull(result.get("data"), "expected data to be non-null");
    }

    if (!setup.live) {
      Map<String, Object> dataMap = Helpers.toMapAny(result.get("data"));
      if (dataMap != null) {
        assertEquals("direct01", dataMap.get("id"), "expected data.id to be direct01");
      }

      assertEquals(1, setup.calls.size(), "expected 1 call");
      Map<String, Object> call = setup.calls.get(0);
      Map<String, Object> initMap = Helpers.toMapAny(call.get("init"));
      if (initMap != null) {
        assertEquals("GET", initMap.get("method"), "expected method GET");
      }
      String url = call.get("url") instanceof String ? (String) call.get("url") : "";
      assertTrue(url.contains("direct01"),
          "expected url to contain direct01, got " + url);
      assertTrue(url.contains("direct02"),
          "expected url to contain direct02, got " + url);
    }
  }

  static class DirectSetup {
    SolardemoSDK client;
    List<Map<String, Object>> calls;
    boolean live;
    Map<String, Object> idmap;
  }

  static DirectSetup directSetup(Object mockres) {
    RunnerSupport.loadEnvLocal();

    final List<Map<String, Object>> calls = new ArrayList<>();

    Map<String, Object> envm = new LinkedHashMap<>();
    envm.put("SOLARDEMO_TEST_MOON_ENTID", new LinkedHashMap<>());
    envm.put("SOLARDEMO_TEST_LIVE", "FALSE");
    Map<String, Object> env = RunnerSupport.envOverride(envm);

    boolean live = "TRUE".equals(env.get("SOLARDEMO_TEST_LIVE"));

    DirectSetup setup = new DirectSetup();
    setup.calls = calls;

    if (live) {
      // sdk-test-control.json's test.client.options seeds the live
      // client; the generated fields below overwrite anything they name.
      Map<String, Object> mergedOpts =
          new LinkedHashMap<>(RunnerSupport.liveClientOptions());
      setup.client = new SolardemoSDK(mergedOpts);
      setup.live = true;

      Map<String, Object> idmap = new LinkedHashMap<>();
      Object entidRaw = env.get("SOLARDEMO_TEST_MOON_ENTID");
      if (entidRaw instanceof String && ((String) entidRaw).startsWith("{")) {
        Map<String, Object> parsed = Helpers.toMapAny(Json.parseOrNull((String) entidRaw));
        if (parsed != null) {
          idmap = parsed;
        }
      }
      else if (entidRaw instanceof Map) {
        idmap = (Map<String, Object>) entidRaw;
      }
      setup.idmap = idmap;
      return setup;
    }

    final Object mockdata = mockres != null ? mockres : jm("id", "direct01");
    BiFunction<String, Map<String, Object>, Map<String, Object>> mockFetch =
        (url, init) -> {
          calls.add(jm("url", url, "init", init));
          return jm(
              "status", 200,
              "statusText", "OK",
              "headers", new LinkedHashMap<>(),
              "json", (Supplier<Object>) () -> mockdata);
        };

    setup.client = new SolardemoSDK(jm(
        "base", "http://localhost:8080",
        "system", jm("fetch", mockFetch)));
    setup.live = false;
    setup.idmap = new LinkedHashMap<>();
    return setup;
  }
}
